rs/pocket_ic_server/src/state_api/state.rs

```diff
@@ -709,12 +709,21 @@ impl ApiState {
         self.stop_progress(instance_id).await;
         let instances = self.instances.read().await;
         let mut canister_http_adapters = self.canister_http_adapters.write().await;
-        canister_http_adapters.remove(&instance_id);
-        let mut instance_state = instances[instance_id].lock().await;
-        if let InstanceState::Available(pocket_ic) =
-            std::mem::replace(&mut *instance_state, InstanceState::Deleted)
-        {
-            std::mem::drop(pocket_ic);
+        loop {
+            let mut instance_state = instances[instance_id].lock().await;
+            match std::mem::replace(&mut *instance_state, InstanceState::Deleted) {
+                InstanceState::Available(pocket_ic) => {
+                    std::mem::drop(pocket_ic);
```

Comment:
Note that the order of dropping `pocket_ic` and the `canister_http_adapters` entry has changed. I don't think it makes a difference though.

Reply:
This is deliberate and I believe it's better to drop the canister HTTP outcall adapter at the very end (in particular, if the instance is still busy, its adapter should not be dropped too early).